Rebuild on changes to the deployment target when compiling Apple targets

madsmtm opened the pull request:
Rebuild, both in Cargo and the incremental cache, when the `*_DEPLOYMENT_TARGET` environment variables change.

Fixes https://github.com/rust-lang/rust/issues/118204. See https://github.com/rust-lang/rust/issues/129432 for more motivation behind this change.

I am a bit unsure if I have implemented this correctly, I'm particularly uncertain whether creating a new field in `Options` the correct approach for busting the incremental cache? **It's only the codegen that needs to be re-run**, everything before that shouldn't be influenced by this, so if there's a way to make that happen, we should do that instead.

Builds upon https://github.com/rust-lang/rust/pull/130068.

r? thomcc

CC @BlackHoleFox
@rustbot blocked

jieyouxu commented:
> I'd also like some input on how to add a test for these kinds of incremental compilation changes?

If the incremental test suite doesn't suit your needs, you could fallback to a run-make test if needed.

madsmtm commented:
> > I'd also like some input on how to add a test for these kinds of incremental compilation changes?
> 
> If the incremental test suite doesn't suit your needs, you could fallback to a run-make test if needed.

Thanks for the tip, I took a look at the incremental test suite, but ended up creating a `run-make` test instead. Note that this change is going to conflict with https://github.com/rust-lang/rust/pull/128507, since I updated the test from in there.

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10490815743/job/29058439222) [(plain)](https://github.com/rust-lang/rust/commit/ed9370bfd8346632b864a00c3ffa65fa1a07f36e/checks/29058439222/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 2.871 Building wheels for collected packages: reuse
#13 2.872   Building wheel for reuse (pyproject.toml): started
#13 3.119   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.120   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132715 sha256=dfa09868353292d98f811d3efdb0d54d07389e808efc71d68e3b93c514bf8bec
#13 3.121   Stored in directory: /tmp/pip-ephem-wheel-cache-lswa5hr2/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 3.123 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.534 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.534 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 4.066 Collecting virtualenv
#13 4.066 Collecting virtualenv
#13 4.116   Downloading virtualenv-20.26.3-py3-none-any.whl (5.7 MB)
#13 4.345      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.7/5.7 MB 25.2 MB/s eta 0:00:00
#13 4.390 Collecting distlib<1,>=0.3.7
#13 4.397   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#13 4.410      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 41.2 MB/s eta 0:00:00
#13 4.449 Collecting filelock<4,>=3.12.2
#13 4.456   Downloading filelock-3.15.4-py3-none-any.whl (16 kB)
#13 4.493 Collecting platformdirs<5,>=3.9.1
#13 4.500   Downloading platformdirs-4.2.2-py3-none-any.whl (18 kB)
#13 4.581 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.771 Successfully installed distlib-0.3.8 filelock-3.15.4 platformdirs-4.2.2 virtualenv-20.26.3
#13 DONE 4.9s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      178112 kB
DirectMap2M:     8210432 kB
DirectMap1G:    10485760 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/59a74db37df8628f778faf75faa41958bb07a42a/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-59a74db37df8628f778faf75faa41958bb07a42a-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
##[endgroup]
fmt check
Diff in /checkout/tests/run-make/apple-deployment-target/rmake.rs:7:
 
 //@ only-apple
-use run_make_support::{cmd, rustc, target, apple_os};
+use run_make_support::{apple_os, cmd, rustc, target};
 
 
 /// Run vtool to check the `minos` field in LC_BUILD_VERSION.
Diff in /checkout/tests/run-make/apple-deployment-target/rmake.rs:14:
Diff in /checkout/tests/run-make/apple-deployment-target/rmake.rs:14:
 /// On lower deployment targets, LC_VERSION_MIN_MACOSX, LC_VERSION_MIN_IPHONEOS and similar
 /// are used instead of LC_BUILD_VERSION - these have a `version` field, so also check that.
 fn minos(file: &str, version: &str) {
-    cmd("vtool").arg("-show-build").arg(file).run().assert_stdout_contains_regex(format!("(minos|version) {version}"));
+    cmd("vtool")
+        .arg("-show-build")
+        .arg(file)
+        .run()
+        .assert_stdout_contains_regex(format!("(minos|version) {version}"));
 
 fn main() {
Diff in /checkout/tests/run-make/apple-deployment-target/rmake.rs:26:
Diff in /checkout/tests/run-make/apple-deployment-target/rmake.rs:26:
         "visionos" => ("XROS_DEPLOYMENT_TARGET", "1.1", "1.2"),
     };
     };
-    let default_version = rustc().target(target()).env_remove(env_var).print("deployment-target").run().stdout_utf8();
+    let default_version =
+        rustc().target(target()).env_remove(env_var).print("deployment-target").run().stdout_utf8();
     let default_version = default_version.strip_prefix("deployment_target=").unwrap().trim();
     // Ensure that the versions we test with are not the default
     // Ensure that the versions we test with are not the default
fmt error: Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/compiler/rustc_trait_selection/src/infer.rs" "/checkout/tests/run-make/rust-lld-by-default-beta-stable/main.rs" "/checkout/tests/run-make/rust-lld-by-default-beta-stable/rmake.rs" "/checkout/tests/run-make/apple-deployment-target/foo.rs" "/checkout/tests/run-make/apple-deployment-target/rmake.rs" "/checkout/tests/run-make/override-aliased-flags/main.rs" "/checkout/tests/run-make/override-aliased-flags/rmake.rs" "/checkout/tests/run-make/CURRENT_RUSTC_VERSION/main.rs" "/checkout/tests/run-make/CURRENT_RUSTC_VERSION/stable.rs" "/checkout/tests/run-make/CURRENT_RUSTC_VERSION/rmake.rs" "/checkout/tests/run-make/rustdoc-output-stdout/foo.rs" "/checkout/tests/run-make/rustdoc-output-stdout/rmake.rs" "/checkout/compiler/rustc_trait_selection/src/traits/vtable.rs" "/checkout/compiler/rustc_trait_selection/src/traits/normalize.rs" "/checkout/compiler/rustc_trait_selection/src/traits/const_evaluatable.rs" "/checkout/compiler/rustc_trait_selection/src/traits/coherence.rs" "/checkout/compiler/rustc_trait_selection/src/traits/mod.rs" "/checkout/compiler/rustc_trait_selection/src/traits/wf.rs" "/checkout/compiler/rustc_trait_selection/src/traits/project.rs" "/checkout/compiler/rustc_trait_selection/src/traits/engine.rs" "/checkout/compiler/rustc_trait_selection/src/traits/auto_trait.rs" "/checkout/compiler/rustc_trait_selection/src/traits/fulfill.rs" "/checkout/compiler/rustc_trait_selection/src/traits/object_safety.rs" "/checkout/tests/run-make/wasm-stringify-ints-small/foo.rs" "/checkout/tests/run-make/wasm-stringify-ints-small/rmake.rs" "/checkout/compiler/rustc_trait_selection/src/traits/specialize/specialization_graph.rs" "/checkout/compiler/rustc_trait_selection/src/traits/specialize/mod.rs" "/checkout/compiler/rustc_trait_selection/src/traits/structural_normalize.rs" "/checkout/compiler/rustc_trait_selection/src/traits/util.rs" "/checkout/compiler/rustc_trait_selection/src/traits/outlives_bounds.rs" "/checkout/tests/run-make/cross-lang-lto/main.rs" "/checkout/tests/run-make/cross-lang-lto/lib.rs" "/checkout/tests/run-make/cross-lang-lto/rmake.rs" "/checkout/compiler/rustc_trait_selection/src/traits/select/candidate_assembly.rs" "/checkout/compiler/rustc_trait_selection/src/traits/select/mod.rs" "/checkout/compiler/rustc_trait_selection/src/traits/select/_match.rs" "/checkout/compiler/rustc_trait_selection/src/traits/select/confirmation.rs" "/checkout/compiler/rustc_trait_selection/src/traits/misc.rs" "/checkout/tests/run-make/lib-trait-for-trait-no-ice/foo.rs" "/checkout/tests/run-make/lib-trait-for-trait-no-ice/rmake.rs" "/checkout/tests/run-make/wasm-custom-sections-opt/foo.rs" "/checkout/tests/run-make/wasm-custom-sections-opt/rmake.rs" "/checkout/compiler/rustc_trait_selection/src/traits/query/normalize.rs" "/checkout/compiler/rustc_trait_selection/src/traits/query/mod.rs" "/checkout/tests/run-make/glibc-staticlib-args/library.rs" "/checkout/tests/run-make/glibc-staticlib-args/rmake.rs" "/checkout/tests/run-make/compiler-rt-works-on-mingw/foo.rs" "/checkout/tests/run-make/compiler-rt-works-on-mingw/rmake.rs" "/checkout/compiler/rustc_trait_selection/src/traits/query/type_op/normalize.rs" "/checkout/compiler/rustc_trait_selection/src/traits/query/type_op/custom.rs" "/checkout/compiler/rustc_trait_selection/src/traits/query/type_op/prove_predicate.rs" "/checkout/compiler/rustc_trait_selection/src/traits/query/type_op/mod.rs" "/checkout/compiler/rustc_trait_selection/src/traits/query/type_op/ascribe_user_type.rs" "/checkout/compiler/rustc_trait_selection/src/traits/query/type_op/outlives.rs" "/checkout/compiler/rustc_trait_selection/src/traits/query/type_op/implied_outlives_bounds.rs" "/checkout/compiler/rustc_trait_selection/src/traits/query/dropck_outlives.rs" "/checkout/compiler/rustc_trait_selection/src/traits/query/evaluate_obligation.rs" "/checkout/compiler/rustc_trait_selection/src/traits/query/method_autoderef.rs" "/checkout/compiler/rustc_trait_selection/src/lib.rs" "/checkout/tests/run-make/use-suggestions-rust-2018/use-suggestions.rs" "/checkout/tests/run-make/use-suggestions-rust-2018/rmake.rs" "/checkout/tests/run-make/use-suggestions-rust-2018/ep-nested-lib.rs" "/checkout/tests/run-make/dylib-soname/foo.rs" "/checkout/compiler/rustc_trait_selection/src/error_reporting/infer/note_and_explain.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Wed Aug 21 13:29:36 UTC 2024
  network time: Wed, 21 Aug 2024 13:29:36 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10491076180/job/29059324945) [(plain)](https://github.com/rust-lang/rust/commit/dc8b9f9d9b684e2cd2937d124b2397f00296adf8/checks/29059324945/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 4.334   Downloading filelock-3.15.4-py3-none-any.whl (16 kB)
#13 4.355 Collecting distlib<1,>=0.3.7
#13 4.370   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#13 4.381      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 50.1 MB/s eta 0:00:00
#13 4.414 Collecting platformdirs<5,>=3.9.1
#13 4.421   Downloading platformdirs-4.2.2-py3-none-any.whl (18 kB)
#13 4.503 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.689 Successfully installed distlib-0.3.8 filelock-3.15.4 platformdirs-4.2.2 virtualenv-20.26.3
#13 DONE 4.8s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      174016 kB
DirectMap2M:     9263104 kB
DirectMap1G:     9437184 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/59a74db37df8628f778faf75faa41958bb07a42a/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-59a74db37df8628f778faf75faa41958bb07a42a-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
##[endgroup]
fmt check
Diff in /checkout/tests/run-make/apple-deployment-target/rmake.rs:7:
 
 //@ only-apple
-use run_make_support::{cmd, rustc, target, apple_os};
+use run_make_support::{apple_os, cmd, rustc, target};
 
 
 /// Run vtool to check the `minos` field in LC_BUILD_VERSION.
Diff in /checkout/tests/run-make/apple-deployment-target/rmake.rs:14:
Diff in /checkout/tests/run-make/apple-deployment-target/rmake.rs:14:
 /// On lower deployment targets, LC_VERSION_MIN_MACOSX, LC_VERSION_MIN_IPHONEOS and similar
 /// are used instead of LC_BUILD_VERSION - these have a `version` field, so also check that.
 fn minos(file: &str, version: &str) {
-    cmd("vtool").arg("-show-build").arg(file).run().assert_stdout_contains_regex(format!("(minos|version) {version}"));
+    cmd("vtool")
+        .arg("-show-build")
+        .arg(file)
+        .run()
+        .assert_stdout_contains_regex(format!("(minos|version) {version}"));
 
 fn main() {
Diff in /checkout/tests/run-make/apple-deployment-target/rmake.rs:26:
Diff in /checkout/tests/run-make/apple-deployment-target/rmake.rs:26:
         "visionos" => ("XROS_DEPLOYMENT_TARGET", "1.1", "1.2"),
     };
     };
-    let default_version = rustc().target(target()).env_remove(env_var).print("deployment-target").run().stdout_utf8();
+    let default_version =
+        rustc().target(target()).env_remove(env_var).print("deployment-target").run().stdout_utf8();
     let default_version = default_version.strip_prefix("deployment_target=").unwrap().trim();
     // Ensure that the versions we test with are not the default
     // Ensure that the versions we test with are not the default
fmt error: Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/tests/run-make/allow-warnings-cmdline-stability/rmake.rs" "/checkout/tests/run-make/mixing-formats/bar1.rs" "/checkout/tests/run-make/mixing-formats/foo.rs" "/checkout/tests/run-make/mixing-formats/rmake.rs" "/checkout/tests/run-make/mixing-formats/baz2.rs" "/checkout/tests/run-make/pgo-gen/test.rs" "/checkout/tests/run-make/pgo-gen/rmake.rs" "/checkout/tests/run-make/archive-duplicate-names/bar.rs" "/checkout/tests/run-make/archive-duplicate-names/foo.rs" "/checkout/tests/run-make/archive-duplicate-names/rmake.rs" "/checkout/tests/run-make/comment-section/rmake.rs" "/checkout/tests/run-make/missing-crate-dependency/crateC.rs" "/checkout/tests/run-make/missing-crate-dependency/crateB.rs" "/checkout/tests/run-make/missing-crate-dependency/crateA.rs" "/checkout/tests/run-make/missing-crate-dependency/rmake.rs" "/checkout/tests/run-make/incr-foreign-head-span/first_crate.rs" "/checkout/tests/run-make/incr-foreign-head-span/second_crate.rs" "/checkout/tests/run-make/incr-foreign-head-span/rmake.rs" "/checkout/tests/run-make/allocator-shim-circular-deps/main.rs" "/checkout/tests/run-make/allocator-shim-circular-deps/my_lib.rs" "/checkout/tests/run-make/allocator-shim-circular-deps/rmake.rs" "/checkout/tests/run-make/rust-lld-custom-target/lib.rs" "/checkout/tests/run-make/rust-lld-custom-target/rmake.rs" "/checkout/tests/run-make/profile/test.rs" "/checkout/tests/run-make/profile/rmake.rs" "/checkout/tests/run-make/dos-device-input/rmake.rs" "/checkout/tests/run-make/rust-lld-by-default-beta-stable/main.rs" "/checkout/tests/run-make/rust-lld-by-default-beta-stable/rmake.rs" "/checkout/tests/run-make/cdylib/bar.rs" "/checkout/tests/run-make/cdylib/foo.rs" "/checkout/tests/run-make/cdylib/rmake.rs" "/checkout/tests/run-make/apple-deployment-target/foo.rs" "/checkout/tests/run-make/apple-deployment-target/rmake.rs" "/checkout/tests/run-make/override-aliased-flags/main.rs" "/checkout/tests/run-make/no-cdylib-as-rdylib/bar.rs" "/checkout/tests/run-make/override-aliased-flags/rmake.rs" "/checkout/tests/run-make/no-cdylib-as-rdylib/foo.rs" "/checkout/tests/run-make/no-cdylib-as-rdylib/rmake.rs" "/checkout/tests/run-make/wasm-spurious-import/main.rs" "/checkout/tests/run-make/wasm-spurious-import/rmake.rs" "/checkout/tests/run-make/CURRENT_RUSTC_VERSION/main.rs" "/checkout/tests/run-make/CURRENT_RUSTC_VERSION/stable.rs" "/checkout/tests/run-make/CURRENT_RUSTC_VERSION/rmake.rs" "/checkout/tests/run-make/rustdoc-output-stdout/foo.rs" "/checkout/tests/run-make/rustdoc-output-stdout/rmake.rs" "/checkout/tests/run-make/intrinsic-unreachable/exit-ret.rs" "/checkout/tests/run-make/intrinsic-unreachable/rmake.rs" "/checkout/tests/run-make/intrinsic-unreachable/exit-unreachable.rs" "/checkout/tests/run-make/wasm-stringify-ints-small/foo.rs" "/checkout/tests/run-make/wasm-stringify-ints-small/rmake.rs" "/checkout/tests/run-make/cross-lang-lto/main.rs" "/checkout/tests/run-make/cross-lang-lto/lib.rs" "/checkout/tests/run-make/cross-lang-lto/rmake.rs" "/checkout/tests/run-make/lib-trait-for-trait-no-ice/foo.rs" "/checkout/tests/run-make/lib-trait-for-trait-no-ice/rmake.rs" "/checkout/tests/run-make/wasm-custom-sections-opt/foo.rs" "/checkout/tests/run-make/wasm-custom-sections-opt/rmake.rs" "/checkout/tests/run-make/glibc-staticlib-args/library.rs" "/checkout/tests/run-make/glibc-staticlib-args/rmake.rs" "/checkout/tests/run-make/compiler-rt-works-on-mingw/foo.rs" "/checkout/tests/run-make/compiler-rt-works-on-mingw/rmake.rs" "/checkout/tests/run-make/use-suggestions-rust-2018/use-suggestions.rs" "/checkout/tests/run-make/use-suggestions-rust-2018/rmake.rs" "/checkout/tests/run-make/mixing-formats/baz.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Wed Aug 21 13:45:49 UTC 2024
  network time: Wed, 21 Aug 2024 13:45:49 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10492751882/job/29065038873) [(plain)](https://github.com/rust-lang/rust/commit/fb6ccd3d2fd5e440a24020c692814f5da3fb5b2c/checks/29065038873/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 2.840 Building wheels for collected packages: reuse
#13 2.841   Building wheel for reuse (pyproject.toml): started
#13 3.085   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.086   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132715 sha256=dfa09868353292d98f811d3efdb0d54d07389e808efc71d68e3b93c514bf8bec
#13 3.086   Stored in directory: /tmp/pip-ephem-wheel-cache-_8z972pe/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 3.089 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.480 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.480 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 4.006 Collecting virtualenv
#13 4.006 Collecting virtualenv
#13 4.060   Downloading virtualenv-20.26.3-py3-none-any.whl (5.7 MB)
#13 4.284      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.7/5.7 MB 25.6 MB/s eta 0:00:00
#13 4.344 Collecting filelock<4,>=3.12.2
#13 4.352   Downloading filelock-3.15.4-py3-none-any.whl (16 kB)
#13 4.372 Collecting distlib<1,>=0.3.7
#13 4.379   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#13 4.390      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 55.7 MB/s eta 0:00:00
#13 4.422 Collecting platformdirs<5,>=3.9.1
#13 4.430   Downloading platformdirs-4.2.2-py3-none-any.whl (18 kB)
#13 4.510 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.694 Successfully installed distlib-0.3.8 filelock-3.15.4 platformdirs-4.2.2 virtualenv-20.26.3
#13 DONE 4.8s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      225216 kB
DirectMap2M:    10260480 kB
DirectMap1G:     8388608 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/59a74db37df8628f778faf75faa41958bb07a42a/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-59a74db37df8628f778faf75faa41958bb07a42a-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
    Finished `release` profile [optimized] target(s) in 28.61s
##[endgroup]
fmt check
fmt: checked 5503 files
Diff in /checkout/tests/run-make/apple-deployment-target/rmake.rs:27:
     let (env_var, example_version, higher_example_version) = match apple_os() {
         "macos" => ("MACOSX_DEPLOYMENT_TARGET", "12.0", "13.0"),
         // armv7s-apple-ios and i386-apple-ios only supports iOS 10.0
-        "ios" if target() == "armv7s-apple-ios" || target() == "i386-apple-ios" => ("IPHONEOS_DEPLOYMENT_TARGET", "10.0", "10.0"),
+        "ios" if target() == "armv7s-apple-ios" || target() == "i386-apple-ios" => {
+            ("IPHONEOS_DEPLOYMENT_TARGET", "10.0", "10.0")
+        }
         "ios" => ("IPHONEOS_DEPLOYMENT_TARGET", "15.0", "16.0"),
         "watchos" => ("WATCHOS_DEPLOYMENT_TARGET", "7.0", "9.0"),
         "tvos" => ("TVOS_DEPLOYMENT_TARGET", "14.0", "15.0"),
fmt error: Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/src/etc/test-float-parse/src/validate/tests.rs" "/checkout/src/etc/test-float-parse/src/traits.rs" "/checkout/src/etc/test-float-parse/src/lib.rs" "/checkout/src/etc/test-float-parse/src/ui.rs" "/checkout/src/etc/test-float-parse/src/validate.rs" "/checkout/src/etc/test-float-parse/src/gen/exponents.rs" "/checkout/src/etc/test-float-parse/src/gen/sparse.rs" "/checkout/src/etc/test-float-parse/src/gen/long_fractions.rs" "/checkout/src/etc/test-float-parse/src/gen/spot_checks.rs" "/checkout/src/etc/test-float-parse/src/gen/exhaustive.rs" "/checkout/src/etc/test-float-parse/src/gen/subnorm.rs" "/checkout/src/etc/test-float-parse/src/gen/integers.rs" "/checkout/src/etc/test-float-parse/src/gen/fuzz.rs" "/checkout/src/etc/test-float-parse/src/gen/many_digits.rs" "/checkout/tests/run-make/box-struct-no-segfault/main.rs" "/checkout/tests/run-make/box-struct-no-segfault/foo.rs" "/checkout/tests/run-make/box-struct-no-segfault/rmake.rs" "/checkout/tests/run-make/env-dep-info/main.rs" "/checkout/tests/run-make/env-dep-info/macro_use.rs" "/checkout/tests/run-make/env-dep-info/rmake.rs" "/checkout/tests/run-make/env-dep-info/macro_def.rs" "/checkout/tests/run-make/pretty-print-with-dep-file/with-dep.rs" "/checkout/tests/run-make/pretty-print-with-dep-file/rmake.rs" "/checkout/tests/run-make/stdin-rustdoc/rmake.rs" "/checkout/tests/run-make/windows-binary-no-external-deps/hello.rs" "/checkout/tests/run-make/windows-binary-no-external-deps/rmake.rs" "/checkout/tests/run-make/sysroot-crates-are-unstable/rmake.rs" "/checkout/tests/run-make/stdin-rustc/rmake.rs" "/checkout/tests/run-make/mixing-formats/bar2.rs" "/checkout/tests/run-make/mixing-formats/baz.rs" "/checkout/tests/run-make/mixing-formats/bar1.rs" "/checkout/tests/run-make/mixing-formats/foo.rs" "/checkout/tests/run-make/mixing-formats/rmake.rs" "/checkout/tests/run-make/mixing-formats/baz2.rs" "/checkout/tests/run-make/archive-duplicate-names/bar.rs" "/checkout/tests/run-make/archive-duplicate-names/foo.rs" "/checkout/tests/run-make/archive-duplicate-names/rmake.rs" "/checkout/tests/run-make/comment-section/rmake.rs" "/checkout/tests/run-make/incr-foreign-head-span/first_crate.rs" "/checkout/tests/run-make/incr-foreign-head-span/second_crate.rs" "/checkout/tests/run-make/incr-foreign-head-span/rmake.rs" "/checkout/tests/run-make/rust-lld-custom-target/lib.rs" "/checkout/tests/run-make/rust-lld-custom-target/rmake.rs" "/checkout/tests/run-make/dos-device-input/rmake.rs" "/checkout/tests/run-make/rust-lld-by-default-beta-stable/main.rs" "/checkout/tests/run-make/rust-lld-by-default-beta-stable/rmake.rs" "/checkout/tests/run-make/apple-deployment-target/foo.rs" "/checkout/tests/run-make/apple-deployment-target/rmake.rs" "/checkout/tests/run-make/override-aliased-flags/main.rs" "/checkout/tests/run-make/override-aliased-flags/rmake.rs" "/checkout/tests/run-make/CURRENT_RUSTC_VERSION/main.rs" "/checkout/tests/run-make/CURRENT_RUSTC_VERSION/stable.rs" "/checkout/tests/run-make/CURRENT_RUSTC_VERSION/rmake.rs" "/checkout/tests/run-make/rustdoc-output-stdout/foo.rs" "/checkout/tests/run-make/rustdoc-output-stdout/rmake.rs" "/checkout/tests/run-make/wasm-stringify-ints-small/foo.rs" "/checkout/tests/run-make/wasm-stringify-ints-small/rmake.rs" "/checkout/tests/run-make/cross-lang-lto/main.rs" "/checkout/tests/run-make/cross-lang-lto/lib.rs" "/checkout/tests/run-make/cross-lang-lto/rmake.rs" "/checkout/tests/run-make/lib-trait-for-trait-no-ice/foo.rs" "/checkout/tests/run-make/lib-trait-for-trait-no-ice/rmake.rs" "/checkout/tests/run-make/wasm-custom-sections-opt/foo.rs" "/checkout/src/etc/test-float-parse/src/main.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Wed Aug 21 15:29:44 UTC 2024
  network time: Wed, 21 Aug 2024 15:29:44 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

madsmtm committed:
Apple: Rebuild when deployment target changes

Rebuild when the `*_DEPLOYMENT_TARGET` variables change when building
Apple targets.

This is done by:
1. Adding it as a tracked variable to `Options` to make sure it clears
   the incremental cache.
2. Emitting the variable in `--emit=dep-info` (`.d`) files, so that
   Cargo can pick up changes to it, and correctly trigger a rebuild.

jieyouxu commented:
> Note that this change is going to conflict with #128507, since I updated the test from in there.

I'll discuss this with @Oneirical, if this PR is updating the macos deployment test we might as well drop the port in https://github.com/rust-lang/rust/pull/128507.

jieyouxu commented on compiler/rustc_driver_impl/src/lib.rs:
```diff
@@ -862,9 +862,9 @@ fn print_crate_info(
-                        .expect("unknown Apple target OS");
-                    println_info!("deployment_target={}", format!("{major}.{minor}"))
+                    let (major, minor, patch) = current_apple_deployment_target(&sess.target);
+                    let patch = if patch != 0 { format!(".{patch}") } else { "".into() };
```
Nit:

```suggestion
                    let patch = if patch != 0 { format!(".{patch}") } else { String::new() };
```

jieyouxu commented on compiler/rustc_interface/src/passes.rs:
```diff
@@ -492,16 +491,22 @@ fn write_out_deps(tcx: TyCtxt<'_>, outputs: &OutputFilenames, out_filenames: &[P
             // Emit special comments with information about accessed environment variables.
             let env_depinfo = sess.psess.env_depinfo.borrow();
+
+            // We will soon sort, so the initial order does not matter.
+            #[allow(rustc::potential_query_instability)]
```
Question [REV1]: this looks a little bit sus. We will sort soon, but the sort below is `sort_unstable`?

jieyouxu commented on compiler/rustc_interface/src/passes.rs:
```diff
@@ -492,16 +491,22 @@ fn write_out_deps(tcx: TyCtxt<'_>, outputs: &OutputFilenames, out_filenames: &[P
+                    (name.into(), std::env::var(name).ok().map(|var| escape_dep_env(&var)))
+                }))
+                .collect();
+            env_depinfo.sort_unstable();
```
Question [REV1]: `sort_unstable` here. Can this ever cause reproducibility problems? Or rather, is there any particular reason this needs to be `sort_unstable` instead of a stable sort?

madsmtm replied:
No idea, this is mostly existing code, I just moved it

madsmtm replied:
Thinking about it, I doubt that the ordering matters, this is environment variable names and values that are passed to Cargo, which in turn uses them all at the same time to determine whether to rebuild.

jieyouxu commented on compiler/rustc_target/src/spec/base/apple/mod.rs:
```diff
@@ -97,53 +98,95 @@ impl TargetAbi {
+    // be safe, and to avoid a warning, we set the architecture explicitly.
+    //
+    // Supported architecture names can be found in the source:
+    // https://github.com/apple-oss-distributions/ld64/blob/ld64-951.9/src/abstraction/MachOFileAbstraction.hpp#L578-L648
```
Remark: thank you for the link, this is very helpful 👍

jieyouxu commented on tests/run-make/apple-deployment-target/rmake.rs:
Remark: these comments are very helpful, thanks!

jieyouxu commented on tests/run-make/apple-deployment-target/rmake.rs:
```diff
@@ -0,0 +1,139 @@
+        rustc.input("foo.rs");
+        rustc.output("foo.o");
+
+        rustc.env(env_var, example_version).run();
```
Remark: not necessarily a problem, but note that all these rustc invocations will dump their output artifacts in the same directory as following invocations. If you want to make sure their output artifacts don't interfere with each other, you can use `run_in_tmpdir`, which will "quarantine" whatever invocations you run within by changing the working directory and output directory of the rustc invocations.

madsmtm replied:
Good point, I've changed the test to use `run_in_tmpdir`

bors commented:
:umbrella: The latest upstream changes (presumably #128507) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

petrochenkov commented:
Blocked on https://github.com/rust-lang/rust/pull/129341.
@rustbot blocked

madsmtm commented:
@rustbot ready

petrochenkov commented:
>Rebuild, both in Cargo and the incremental cache

I think the correct way to do this would be to integrate env var reading into rustc's query system.
We can make an `eval_always` query that reads the variable, then it will automatically poison all its dependent queries if the env var changes between compilations.
This query would then be used for reading *all* environment variables, not just `*_DEPLOYMENT_TARGET`.

The query can also put the read variable into `sess.psess.env_depinfo`, then it will be automatically tracked for cargo as well.

It would also be better to split the Apple-specific parts, including tests, from rustc infrastructure changes, put them into a different PR and land it separately.

petrochenkov commented:
There are some issues related to target specs though.

First, the query system is not available in `rustc_target` when targets are loaded.
So we may need to re-query the relevant env vars used from there later.

Second, with `--check-cfg` we load *all* the targets, including the Apple ones, and depend on the Apple env vars.
`--check-cfg` is enabled by default now, so we technically depend on the Apple-related env vars on all hosts and targets, in most scenarios, if we consider just constructing a `Target(Option)` value a dependency.

Which fields of the target specs actually depend on the env vars?
Maybe the logic related to those fields can be excluded from target specs and moved to some other place (link.rs or where else they are necessary)?
Then the dependencies will be introduced only when strictly necessary.

petrochenkov commented:
Linker arguments depending on env vars can certainly be added in link.rs, instead of target specs, likely in `fn add_order_independent_options`.

The `llvm_target` field can be versionless in target specs, the version can be added only when strictly necessary, e.g. before passing to linker or maybe to LLVM backend as well.

(Maybe I missed some other fields, didn't look very carefully.)

madsmtm commented:
> It would also be better to split the Apple-specific parts, including tests, from rustc infrastructure changes, put them into a different PR and land it separately.

Good point, I've started with the tests in https://github.com/rust-lang/rust/pull/130068, will get back to your other comments.

madsmtm commented:
> Linker arguments depending on env vars can certainly be added in link.rs, instead of target specs, likely in `fn add_order_independent_options`.

Yes, that's also likely what I will end up doing anyhow to implement proper `SDKROOT` reading.

> The `llvm_target` field can be versionless in target specs, the version can be added only when strictly necessary, e.g. before passing to linker or maybe to LLVM backend as well.

I... never actually really considered that, might be doable? And would be a lot cleaner, I really dislike that `rustc_target` is impure in this fashion.

> (Maybe I missed some other fields, didn't look very carefully.)

I don't think you missed any, but I think the question should rather be "which fields could we possible _want_ to depend on the deployment target in the future". That is, I _would_ like to change `rustc_target` so that it doesn't depend on these, but it has to be forwards compatible, so that future people (likely myself) that wants to configure a target specially depending on the deployment target, is empowered to do so.

I'm going through all the current `Target[Options]`, and considering whether they might want to change depending on deployment target:

<details><summary>Details</summary>
<p>

```
// Target
llvm_target -> yes, as discussed
metadata -> no
pointer_width -> no
arch -> possibly, but probably no
data_layout -> possibly

// TargetOptions
is_builtin -> no
endian -> no
c_int_width -> no
os -> no
env -> no
abi -> no
vendor -> no
linker -> no, though might want to depend on the SDK version
linker_flavor -> no
linker_flavor_json -> no
lld_flavor_json -> no
linker_is_gnu_json -> no
*link* -> yes, as discussed, but only desired to load when actually linking
asm_args -> yes
cpu -> yes, definitely
features -> yes, definitely
direct_access_external_data -> no
dynamic_linking -> possibly, watchOS might start supporting it
dll_tls_export -> no
only_cdylib -> no
executables -> no
relocation_model -> unlikely
code_model -> no
tls_model -> no
disable_redzone -> no
frame_pointer -> possibly?
function_sections -> no
dll_prefix -> no
dll_suffix -> no
exe_suffix -> no
staticlib_prefix -> no
staticlib_suffix -> no
families -> unlikely
abi_return_struct_as_int -> no
is_like_* -> no
default_dwarf_version -> yes
allows_weak_linkage -> no
has_rpath -> no
no_default_libraries -> no
position_independent_executables -> no
static_position_independent_executables -> no
plt_by_default -> no
relro_level -> no
archive_format -> no
allow_asm -> no
main_needs_argc_argv -> no
has_thread_local -> would've been useful in the past
obj_is_bitcode -> no
forces_embed_bitcode -> would've been useful in the past
bitcode_llvm_cmdline -> no
min_atomic_width -> no
max_atomic_width -> no
atomic_cas -> no
panic_strategy -> no
crt_static_allows_dylibs -> no
crt_static_default -> no
crt_static_respected -> no
stack_probes -> possibly
min_global_align -> no
default_codegen_units -> no
default_codegen_backend -> no
trap_unreachable -> no
requires_lto -> no
singlethread -> no
no_builtins -> no
default_hidden_visibility -> no
emit_debug_gdb_scripts -> no
requires_uwtable -> no
default_uwtable -> no
simd_types_indirect -> no
limit_rdylib_exports -> no
override_export_symbols -> no
merge_functions -> no
mcount -> no
llvm_mcount_intrinsic -> no
llvm_abiname -> no
relax_elf_relocations -> no
llvm_args -> possibly
use_ctors_section -> no
eh_frame_header -> no
has_thumb_interworking -> no
debuginfo_kind -> no
split_debuginfo -> no
supported_split_debuginfo -> no
supported_sanitizers -> no
c_enum_min_bits -> no
generate_arange_section -> no
supports_stack_protector -> no
entry_name -> no
entry_abi -> no
supports_xray -> no
```

</p>
</details> 

In summary, most of these are really just codegen, though some of them would be interesting for the standard library to do version detection, and emit more performant implementations if the deployment target is high enough. And in those cases, we _do_ want it to be integrated properly with the query system, such that only e.g. the MIR in the standard library that depend on those would need to be changed.

A query like `deployment_target` would also be useful for something like https://github.com/rust-lang/rfcs/pull/3379, if we had `cfg!(os_version_min(...))` we'd want incremental compilation to work with that.

All of this to say that I've convinced myself that turning it into a query is a good idea.

madsmtm commented:
Just to make sure I understand correctly, I would add a `query env_var` in `compiler/rustc_middle/src/query/mod.rs`? And then in another PR I would change `llvm_target` to be versionless in target specs, and make the deployment target be loaded using the query instead? Should that also be a query like `query deployment_target`, or should that just be a regular method that calls `env_var` internally?

And since I'm asking anyways, though I'll do it separately for sure, as part of https://github.com/rust-lang/rust/issues/129432 I know that I'm also gonna need the ability to mark certain system file paths as part of the incremental cache - should this be done generally as well (`query file_changed`), or is it niche enough that it should just be its own `query sdkroot`?

madsmtm commented:
> Linker arguments depending on env vars can certainly be added in link.rs, instead of target specs, likely in `fn add_order_independent_options`.

Actually, how would I go about doing that, if I wanted the deployment target to be a query? It seems like the `TyCtx<'_>` no longer exists when `add_order_independent_options` is called? Although there is `ParseSess::env_depinfo`, do you mean I should use that?

petrochenkov commented:
>Just to make sure I understand correctly, I would add a query env_var in compiler/rustc_middle/src/query/mod.rs?

Yes, seems about right.

>And then in another PR I would change llvm_target to be versionless in target specs, and make the deployment target be loaded using the query instead?

Yep.

>Should that also be a query like query deployment_target, or should that just be a regular method that calls env_var internally?

Just `env_var` query should be enough for a start.

>I'm also gonna need the ability to mark certain system file paths as part of the incremental cache

This is better left for the last, I don't think we are currently doing anything like that in the compiler.

>It seems like the `TyCtx<'_>` no longer exists when `add_order_independent_options` is called?

`add_order_independent_options` has the `codegen_results.crate_info` in arguments, `tcx` is available when it is constructed.

petrochenkov commented:
https://github.com/rust-lang/rust/pull/130068 has landed, unblocking.
@rustbot author

madsmtm commented:
> > It seems like the `TyCtx<'_>` no longer exists when `add_order_independent_options` is called?
> 
> `add_order_independent_options` has the `codegen_results.crate_info` in arguments, `tcx` is available when it is constructed.

Ah, I see now, thanks!

I will try to find some time next week to implement what we've discussed.

bors commented:
:umbrella: The latest upstream changes (presumably #132497) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Dylan-DPC commented:
@madsmtm any updates on this? thanks

petrochenkov commented:
https://github.com/rust-lang/rust/pull/130883 has landed, unblocking.
@rustbot author

Dylan-DPC commented:
@madsmtm this is no longer blocked so any updates on this?
